cas: Introduce db handler (based on ql) and rewrite remote to use it.

sgotti opened the pull request:
This PR tries to introduce a sql like db instead of diskv to save the additional indexes (other than the ACI contents). By now remote is converted to a sql db table. I'm going to try converting also aciinfo proposed in #352.

cas/db.go introduces a db handler and by now uses [ql](github.com/cznic/ql) but it can be replaced by another database (like sqlite).

ql is an embedded database that should be opened only by one process to avoid data corruption. Internally it already locks the file but if the lock is already held by another process the Open will fail. Instead other rkt processes should wait for the lock to be freed.
This is done locking the dir containing the db file.

To avoid long wait times of concurrent rkt processes trying to access a DB, only short living transactions should be done.
The easiest path is to open and close the db for every transaction. This will trigger a possible db log replay (but it should be fast if the db has been cleanly closed) and can sound as a wrong pattern but it will avoid a lot of troubles like dealing with transaction isolation levels (if multiple concurrent transaction were used) etc..

Sqlite supports multiple concurrent processes accessing the same db file but perhaps, if the above pattern works well, it should't be needed (or it will create other problems...).

The DoTx function executes the provided txfuncs inside a unique transaction.
If one of the functions returns an error the whole transaction is rolled back.
The Do functions calls DoTx but before it locks and opens the db, and after it closes and unlocks the db.
For the reasons explained above the Do function should always be used.

sgotti commented:
I updated the patch after the signature validation work introduced by #335.

philips commented:
Great! Thanks for taking a swing at this and sorry for the delay reviewing.

philips commented:
Overall this LGTM. I would like to use sqlite instead of ql because of the multi-process reader thing though. It will likely require making our build more complicated though.

sgotti commented:
@philips thanks for your review. I updated the PR with your suggestions. I was already thinking about a version table and migration paths but I was waiting for an initial review. By now I left the migration code for another PR that I'll open when this is merged.

I also removed the scheme (added by the docker2aci integration) to the remote type as I don't think it's useful to save it in a db column.

About the ql vs sqlite: I completely avoided the multi process writers and readers with the actual lock/open/close/unlock pattern to avoid possible problems due to concurrent transactions (but sqlite uses serializable transactions isolation so it shouldn't be a problem). Are you asking to remove the locking/open/close/unlock code and let sqlite completly handle the locking? Doing this will need a sort of cas.Close() to cleanly close the db on exit and change all the cas users to call it on exit.

philips commented:
LGTM, overall. A second set of eyes would be great. /cc @eyakubovich @vcaputo

sgotti commented:
@philips Updated. Thanks!

philips commented:
/cc @eyakubovich @jonboulle @alban @iaguis @vcaputo Can someone else take a look?

iaguis commented:
LGTM

sgotti commented:
@iaguis Thanks for your review! I updated the PR with your comments.

jonboulle commented:
A few little areas for cleanup (like docstrings) but let's follow them up rather than going through another rebase cycle.

@sgotti thanks for the patch and your patience, this is great!

jonboulle commented:
:+1:
